src/Container/Tarball.hs

```diff
@@ -0,0 +1,183 @@
+          Left err -> Left err
+          Right layer -> Right layer
+
+      -- Layer tarball must be a file, they cannot be a directory
+      -- or symlink to some other file.
+      _ -> Left $ TarLayerNotAFile layerTarball
+
+mkLayerFromOffset ::
+  TarEntryOffset ->
+  Tar.Entries Tar.FormatError ->
+  Either ContainerImgParsingError ContainerLayer
+mkLayerFromOffset imgOffset = build (ContainerLayer mempty 0)
+  where
+    build !builder (Tar.Next e es) = build (addNextChangeSet imgOffset e builder) es
+    build !builder Tar.Done = Right builder
+    build !_ (Tar.Fail err) = Left $ TarParserError err
+
+    addNextChangeSet :: TarEntryOffset -> Tar.Entry -> ContainerLayer -> ContainerLayer
+    addNextChangeSet offset entry containerLayer =
+      ContainerLayer
+        { layerChangeSets = updateChangeSet offset entry containerLayer
+        , lastOffset = nextEntryOffset entry (lastOffset containerLayer)
+        }
+
+    updateChangeSet :: TarEntryOffset -> Tar.Entry -> ContainerLayer -> Seq ContainerFSChangeSet
+    updateChangeSet offset entry containerLayer =
+      if isDoubleWhiteOut (entryTarPath entry)
+        || ( not (isFileOrSymLink entry)
+              && not (isWhiteOut $ entryTarPath entry)
+           )
+        then -- Do not capture Insert for non-files or non-symbolic links, as folders
+        -- by themselves are not analysis relevant, and filepath information already contains
+        -- relevant folder information.
+          layerChangeSets containerLayer
+        else
+          (layerChangeSets containerLayer)
+            |> (mkChangeSet (entryTarPath entry) (offset + lastOffset containerLayer + 1))
+
+    mkChangeSet :: TarPath -> TarEntryOffset -> ContainerFSChangeSet
+    mkChangeSet path offset = case removeWhiteOut $ fromTarPath path of
+      Nothing -> InsertOrUpdate (fromTarPath path) offset
+      Just p -> Whiteout p
+
+-- | True if tar entry is for a file or a symlink, otherwise False
+isFileOrSymLink :: Tar.Entry -> Bool
+isFileOrSymLink e = isFile e || isSymLink e
+
+-- | True if tar entry is for a file with content, otherwise False.
+isFile :: Tar.Entry -> Bool
+isFile (TarEntry.Entry _ (NormalFile _ _) _ _ _ _) = True
+isFile _ = False
+
+-- | True if tar entry is for a symbolic link, otherwise False.
+isSymLink :: Tar.Entry -> Bool
+isSymLink (TarEntry.Entry _ (SymbolicLink _) _ _ _ _) = True
+isSymLink _ = False
+
+-- | True if tar path has double whiteout marker.
+isDoubleWhiteOut :: TarPath -> Bool
+isDoubleWhiteOut = fileNameHasPrefix ".wh..wh."
+
+-- | True if tar path has whiteout marker.
+isWhiteOut :: TarPath -> Bool
+isWhiteOut = fileNameHasPrefix ".wh"
+
+-- | True if tarpath's filename has provided prefix. Otherwise False.
+fileNameHasPrefix :: Text -> TarPath -> Bool
+fileNameHasPrefix prefix path = Text.isPrefixOf prefix (fileNameOf $ toText . fromTarPath $ path)
+
+-- | Filename from text path.
+fileNameOf :: Text -> Text
+fileNameOf path = snd $ Text.breakOnEnd "/" path
```

Comment:
Do we need to support Windows here? Would this fail if the paths there use "\\"?

Reply:
We only need to support POSIX, will add comment reflecting to this fact;